primedev/core/filesystem/rpakfilesystem.cpp

```diff
@@ -2,206 +2,424 @@
+			return;
 		}
 	}
 }
 
-// clang-format off
-HOOK(LoadPakAsyncHook, LoadPakAsync,
-int, __fastcall, (char* pPath, void* unknownSingleton, int flags, void* pCallback0, void* pCallback1))
-// clang-format on
+/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+static bool (*o_pLoadMapRpaks)(char* mapPath) = nullptr;
+static bool h_LoadMapRpaks(char* mapPath)
 {
-	HandlePakAliases(&pPath);
+	// unload all mod rpaks that are marked for unload
+	g_pPakLoadManager->UnloadMarkedPaks();
+	g_pPakLoadManager->CleanUpUnloadedPaks();
+
+	// strip file extension
+	const std::string mapName = fs::path(mapPath).replace_extension().string();
+
+	// load mp_common, sp_common etc.
+	o_pLoadGametypeSpecificRpaks(mapName.c_str());
+
+	// unload old modded map paks
+	g_pPakLoadManager->UnloadModPaks();
+	// load modded map paks
+	g_pPakLoadManager->LoadModPaksForMap(mapName.c_str());
 
-	// dont load the pak if it's currently loaded already
-	size_t nPathHash = STR_HASH(pPath);
-	if (g_pPakLoadManager->GetPakHandle(nPathHash) != -1)
-		return -1;
+	// don't load/unload anything when going to the lobby, presumably to save load times when going back to the same map
+	if (!g_pPakLoadManager->GetForceReloadOnMapLoad() && !strcmp("mp_lobby", mapName.c_str()))
+		return false;
 
-	bool bNeedToFreePakName = false;
+	char mapRpakStr[272];
+	snprintf(mapRpakStr, 272, "%s.rpak", mapName.c_str());
 
-	static bool bShouldLoadPaks = true;
-	if (bShouldLoadPaks)
+	// if level being loaded is the same as current level, do nothing
+	if (!g_pPakLoadManager->GetForceReloadOnMapLoad() && !strcmp(mapRpakStr, currentMapRpakPath))
+		return true;
+
+	strcpy(currentMapRpakPath, mapRpakStr);
+
+	(*o_pcleanMaterialSystemStuff)();
+	o_ploadlevelLoadscreen(mapName.c_str());
+
+	// unload old map rpaks
+	int curHandle = *currentMapRpakHandle;
+	int curPatchHandle = *currentMapPatchRpakHandle;
+	if (curHandle != -1)
 	{
-		// make a copy of the path for comparing to determine whether we should load this pak on dedi, before it could get overwritten by
-		// LoadCustomMapPaks
-		std::string originalPath(pPath);
+		(*o_pCModelLoader_UnreferenceAllModels)(*modelLoader);
+		(*o_pcleanMaterialSystemStuff)();
+		g_pakLoadApi->UnloadPak(curHandle, *o_pcleanMaterialSystemStuff);
+		*currentMapRpakHandle = -1;
+	}
+	if (curPatchHandle != -1)
+	{
+		(*o_pCModelLoader_UnreferenceAllModels)(*modelLoader);
+		(*o_pcleanMaterialSystemStuff)();
+		g_pakLoadApi->UnloadPak(curPatchHandle, *o_pcleanMaterialSystemStuff);
+		*currentMapPatchRpakHandle = -1;
+	}
 
-		// disable preloading while we're doing this
-		bShouldLoadPaks = false;
+	*currentMapRpakHandle = g_pakLoadApi->LoadRpakFileAsync(mapRpakStr, *rpakMemoryAllocator, 7);
 
-		LoadPreloadPaks();
-		LoadCustomMapPaks(&pPath, &bNeedToFreePakName);
+	// load special _patch rpak (seemingly used for dev things?)
+	char levelPatchRpakStr[272];
```

Comment:
Where'd you get `272`, might be worth using `std::string` though i dont remember if we have helpers for that

Reply:
On second thought since we end up writing into vanilla memory it's probably best to keep the length fixed and known to match the space that vanilla has?